Design.md

```diff
@@ -0,0 +1,55 @@
+
+Could be improved by sending `SIGTERM` shortly before `SIGKILL`, but I'm aiming for simplicity here. Also separating soft and hard kills in the API level might be useful in some situations.
+
+### Status
+
+Returns job status, i.e. is it running, and the status code if the job has completed. If the job has been terminated with a signal, that is reported instead.
+
+### Output
+
+Streams output of a job in binary blobs. Each blob is tagged to be either from stdout or stderr. Stream is automatically closed when the process completes and all output has been streamed. All calls to output stream the whole output history from the moment the process was started.
+
+All output of each process is stored in a separate memory buffer. For each call of output, a separate offset value for both IO streams is maintained.
+
+## Scalability and high availability
+
+The system is reasonably perfomant, it should be able to handle quite a lot of processes on a single machine. However, there is no way to run this on multiple machines without some coordinator in front. The limits the horizontal scalability to a single machine. Of course it's possible for the client to handle multiple machines, but this is not built into the system. Ideally there would be a way to move processes from machine to another, but whether that's viable depends on many factors. This requires some way to synchronize shared state, usually using a database
+
+As for high availablity, similar concerns apply. Again, coordination with multiple controller nodes might be the best way to solve this. Many things depend on what properties are required. For instance, usually the platform itself cannot handle all high-availability concerns, as those must be solved by the application. However, things like "there should always be at least one copy of this software running" can usually be solved by duplication.
+
+### Other tradeoffs and simplifications
+
+Full output history of all jobs is stored in memory, and is only removed on server restart. In a real system, the output would usually be streamed to a log database, or just into a file, to reduce memory pressure. After process termination logs should be either removed or moved to an archive (e.g. Amazon S3).
+
+Other simplification is about how jobs are identified. A proper implementation would use either UUIDs or user-given names, maybe even both. This implementation, however, uses sequential unsigned 64-bit integers to identify jobs. The CLI tool uses them as well, making it particularly error-prone when jobs are manually referenced. The job ids also reset to zero every time the server is restarted, so that no persitent storage is required. This may cause conflicts if the server is restarted. Using UUIDs would solve this issue automatically.
```

Comment:
other than conflicts, are there any other downsides to using sequential IDs?
also, using UUID in this challenge should be fine, as long as it's a simple dependency call.

Reply:
Sure, I'll use UUIDs if that seems justified. The most important other downsides of sequential include:
* Easier to guess for potential attackers
* Distributed no-coordination creation is impossible